HPCC-18040 Provides unicode implementations for excludeNthWord

dskaff opened the pull request:
- Adds interface definition to ecllibrary/std/uni.ecl
- Adds test cases to mirror string version of excludeNthWord
- Adds code that executes excludeNthWord with unicode capabilities

Signed-off-by: David Skaff <David.Skaff@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 26f536f3824dc109338f864ff83b4ac4a1d99753
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.586 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |742 |742 |0 |
|test (thor) |637 |637 |0 |
|test (roxie) |770 |770 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on ecllibrary/teststd/uni/TestExcludeNthWord.ecl:
```diff
@@ -0,0 +1,65 @@
+     //Test latin diacritical marks
+     EXPORT Test26 := ASSERT(Uni.ExcludeNthWord(U'À à',2)+U'!' = U'À !');
+     EXPORT Test27 := ASSERT(Uni.ExcludeNthWord(U'ä̰́ Ä̰́',2)+U'!' = U'ä̰́ !');
+     //Test Chinese. "I am a computer" --> "I am"
```
A good set of tests.  I would suggest a couple more:
commas and other punctuation: 'abc, def, ghi, jhi'      'abc := name'
diacritics within words

ghalliday replied:
Also, it would be worth including tests that include numbers - both as separate words and combined e.g., "abc 123 abc123 123abc"

- Check that words extracted at all word positions are correct/consistent (I suspect this may not be the case).

ghalliday replied:
Having looked at the string version I see that it only treats characters < 0x20 as break characters.  This is probably a point where the string version and the unicode version diverge.  @johnholt what would you expect?

ghalliday replied:
You should also add similar tests to teststd/uni/TestGetNthWord to ensure the results are consistent.  (The tests in that file should have included tests similar to yours above.)

ghalliday commented on ecllibrary/teststd/uni/TestExcludeNthWord.ecl:
```diff
@@ -0,0 +1,65 @@
+     //Test Chinese. "I am a computer" --> "I am"
+     EXPORT Test28 := ASSERT(Uni.ExcludeNthWord(U'我是電腦',2)+U'!' = U'我是!');
+     /*
+     //arabic lettering with accent marks. Bidirectional.
```
You should add a comment to explain why these tests are commented out.

ghalliday commented on ecllibrary/teststd/uni/TestExcludeNthWord.ecl:
```diff
@@ -0,0 +1,65 @@
+﻿/*##############################################################################
+ 
```
whitespace: this line and the other blank lines in the comment have trailing spaces.  (git diff HEAD^ will highlight them)

ghalliday commented on ecllibrary/teststd/uni/TestExcludeNthWord.ecl:
```diff
@@ -0,0 +1,65 @@
+﻿/*##############################################################################
+ 
+     HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
minor: new files should update the date

ghalliday commented on ecllibrary/std/Uni.ecl:
```diff
@@ -382,4 +382,20 @@ EXPORT unsigned4 WordCount(unicode text, varstring localename = '') :=
+ * Lacks bidirectional capability.
+ *
+ * @param text          The string to be broken into words.
+ * @param n             Which word should be returned from the function.
```
removed from the string

ghalliday commented on ecllibrary/teststd/uni/TestExcludeNthWord.ecl:
```diff
@@ -0,0 +1,65 @@
+     EXPORT Test29 := ASSERT(Uni.ExcludeNthWord(U'حركات إعجام', 2)+U'!' = U' حركات!');
+     //hebrew lettering with accent marks (called pointing). Bidirectional.
+     EXPORT Test30 := ASSERT(Uni.ExcludeNthWord(U'קָמָץ שִׁי״ן',2)+U'!' = U' קָמָץ!');
+     */
```
Are there any locales that make a difference to the word rules?  As far as I can see the special rules for thai/chinese are used whatever the locale.  If there are no examples it would be worth adding a comment to that effect here.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,48 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
 
+UnicodeString excludeNthWord(RuleBasedBreakIterator& bi, UnicodeString const & source, unsigned n)
+{
+	UnicodeString processed;
```
whitespace: This file contains tabs to indent - our standard is to use 4 spaces.

ghalliday replied:
Efficiency: Think about the number of UnicodeString that are created - each one will clone the string on the heap.  I think 3 UnicodeStrings are created, and it is currently likely to clone the string 2 or 3 times, when 1 could be done instead.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,48 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+					wordBeginning = 0;
+				}
+				else {
+					wordBeginning = bi.preceding(idx);
```
Efficiency: Can you avoid doing this by saving values from previous iterations?

dskaff replied:
By this do you mean set wordBeginning's value to the idx before the nth word? @ghalliday

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -1416,3 +1459,30 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleGetNthWord(unsigned & tgtLen,
+{
+    UErrorCode status = U_ZERO_ERROR;
+    Locale locale(localename);
+    RuleBasedBreakIterator* bi = (RuleBasedBreakIterator*)RuleBasedBreakIterator::createWordInstance(locale, status);
```
Efficiency: Note this section in the ICU documentation: http://userguide.icu-project.org/boundaryanalysis
"Reuse

It is slow and wasteful to repeatedly create and destroy a BreakIterator when it is not necessary. For example, do not create a separate BreakIterator for each line in a document that is being word-wrapped. Keep around a single instance of a line BreakIterator and use it whenever a line break iterator is needed.
"
I know you based this on an existing function.  That was written 6 years ago before we had a good code review system in place.  I think it might be best to leave it as it is, but fix it with a separate pull request.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -1416,3 +1459,30 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleGetNthWord(unsigned & tgtLen,
+    UnicodeString uText(text, textLen);
+    UnicodeString processed = excludeNthWord(*bi, uText, n);
+    delete bi;
+	if(processed.length()>0)
```
indent: extra indentation (because of a tab character)

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -1416,3 +1459,30 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleGetNthWord(unsigned & tgtLen,
+        tgt = 0;
+    }
+}
+
```
whitespace: extra whitespace at the end of the file.  Both githooks and "git diff HEAD^ --check" can be used to find the problem.

ghalliday reviewed (changes requested):
@dskaff a good first go.  There are various comments from minor to potentially more significant.

If the effect of this change is to highlight ways of improving getNthWord as well, then that will be useful - particularly since that function is more likely to be used.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 75c859d9cf1890ab790bcd4d6872736dc1952cc6
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.72 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |743 |743 |0 |
|test (thor) |638 |638 |0 |
|test (roxie) |771 |771 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

dskaff commented:
@ghalliday please code review the latest commit.

ghalliday commented on ecllibrary/std/Uni.ecl:
```diff
@@ -382,4 +382,19 @@ EXPORT unsigned4 WordCount(unicode text, varstring localename = '') :=
     lib_unicodelib.UnicodeLib.UnicodeLocaleGetNthWord(text, n, localename);
 
+/**
+ * Returns everything except the nth word and, depending on the nth word, some whitespaces from the string.  Words are marked by the unicode break semantics.
```
minor: We generally aim to have lines less than 120 characters.  Not always enforced, but it would be worth reformatting this so the long lines were a bit shorter.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,43 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+    int32_t idx = bi.first();
+    int32_t wordidx = 0;
+    unsigned wordBeginning = 0;
+    while (idx != BreakIterator::DONE) {
```
style:  The general style is for curly brackets to be on the following line.  The function you started from doesn't follow that style.   So while, if, do in this function should be on a new line.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,43 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+UnicodeString excludeNthWord(RuleBasedBreakIterator& bi, unsigned textLen, UChar const * text, unsigned n)
```
This is an improvement, but the return is likely to involve an extra clone of the string - unless the c++ compiler is very clever.

You could do the following instead:

```
void excludeNthWord(RuleBasedBreakIterator& bi, UnicodeString & source, unsigned n)
...
                source.removeBetween(wordBeginning, wordEnd);
                return;
...
    UnicodeString processed(text, textLen);
    excludeNthWord(*bi, processed, n);

```

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,43 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+        }
+        wordBeginning = idx;
+        idx = bi.next();
+        if (idx == BreakIterator::DONE && wordidx < 1) {
```
I think you can simplify the code by moving these two conditions outside the loop.

ghalliday reviewed (changes requested):
@dskaff looking good.  The test cases are very good.

A few minor layout issues, and a couple of potential efficiency improvements.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 1d7a606699e39e684d43f0222dac8f421a726ad5
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.618 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |743 |743 |0 |
|test (thor) |638 |638 |0 |
|test (roxie) |771 |771 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 21dc283b86aa7c9ae78b9b4c836554c35e6c99a3
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.189 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |743 |743 |0 |
|test (thor) |638 |638 |0 |
|test (roxie) |771 |771 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

dskaff commented:
@ghalliday Should be ready for merge.

dskaff commented:
Edited commit message to not contain the messages of several previous commits. (my fault for not editing it when merging them together)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 62f8ab282245d95504502eb95eb2ed64852d3a19
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.118 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |743 |743 |0 |
|test (thor) |638 |638 |0 |
|test (roxie) |771 |771 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,46 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+    bi.setText(source);
+    int32_t idx = bi.first();
+    int32_t wordidx = 0;
+    unsigned wordBeginning;
```
trivial.  It would be worth assigning 0 to wordBeginning.  The value will never actually be used, but it will avoid some false positives from programs like Coverity which check the correctness of the program - since there is a path through the function where wordBeginning is not initialised (even though the path can never be taken).

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -718,6 +719,46 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+    {
+        source.removeBetween(bi.first(), bi.last());
+    }
+    return;
```
trivial: No need for a return a the end of a function.  Please remove.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -1416,3 +1457,23 @@ UNICODELIB_API void UNICODELIB_CALL ulUnicodeLocaleGetNthWord(unsigned & tgtLen,
+    UnicodeString processed(text, textLen);
+    excludeNthWord(*bi, processed, n);
+    delete bi;
+    if(processed.length()>0)
```
trivial: Normal to have a space between the if and the bracket.

ghalliday reviewed (approved):
@dskaff 3 very trivial comments, otherwise ready to merge.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: a583bd04db65a80e47583f27341708e01e6e71d3
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.069 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

Commit:
HPCC-18040 Provides unicode implementations for excludeNthWord

- Adds interface definition to ecllibrary/std/uni.ecl
- Adds test cases to mirror string version but with Unicode capability
- Adds code that executes excludeNthWord with unicode capabilities

Signed-off-by: David Skaff <David.Skaff@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 5089651b2dae89cbeafd75188754bdfbab71991f
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.431 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |95 |95 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
